### PR TITLE
Add capabilities for storing and replaying a transaction

### PR DESCRIPTION
**Pull request by kontura**
This is a second iteration for: https://github.com/rpm-software-management/dnf5/pull/978 I have made a new PR because it nearly completely changed and the old comments could be confusing.

Now there is only one approach and replaying a transaction must go through a `libdnf5::Goal`. The new API is:
- `void libdnf5::Goal::add_transaction_replay(const std::string & json_serialized_transaction, libdnf5::GoalJobSettings & settings_p);`
- `std::string libdnf5::base::transaction::serialize_to_json();`
- `std::string libdnf5::transaction::transaction::serialize_to_json();`

For: https://github.com/rpm-software-management/dnf5/issues/999

### CODE REVIEW

File: dnf5/commands/history/history_store.cpp [hunk 1]
```diff
@@ -19,14 +19,58 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
...
+    if (transactions.size() != 1) {
+        throw libdnf5::cli::CommandExitError(1, M_("Multiple transactions selected for storing, only one allowed."));
+    }
+
+    std::filesystem::path path(output_option->get_value());
+    std::ofstream serialized_trans;
+    //TODO(amatej): add warning if file already exists?
+    serialized_trans.open(path);
+    serialized_trans << transactions[0].serialize_to_json();
+    serialized_trans.close();
```
Comment (j-mracek) [inline dnf5/commands/history/history_store.cpp:73]: I would recommend to perform the operation atomically .
1. Create a temp file
2. Generate content an save it in temp unique file
3. Move temp file to the new location

This approach will ensure that the output will be 100% correct and not only partially written or corrupted.

File: include/libdnf5/base/goal_elements.hpp [hunk 2]
```diff
@@ -166,9 +166,12 @@ struct GoalJobSettings : public ResolveSpecSettings {
         return group_package_types ? &group_package_types.value() : nullptr;
     }
 
-    /// If set to true, group operations (install / remove) will only work
+    /// If set to true, group operations (install / remove / upgrade) will only work
     /// with the group itself, but will not add to the transaction any packages.
     bool group_no_packages{false};
+    /// If set to true, environment operations (install / remove / upgrade) will only work
+    /// with the environment itself, but will not add to the transaction any groups.
+    bool environment_no_groups{false};
```
Comment (j-mracek) [inline include/libdnf5/base/goal_elements.hpp:174]: This change will brake ABI compatibility (C++).

Comment (ppisar) [reply include/libdnf5/base/goal_elements.hpp:174]: I agree that we should add a pimpl member to every public class on the first need of breaking ABI. So that we can survive without breaking ABI later.

Comment (kontura) [reply include/libdnf5/base/goal_elements.hpp:174]: I am also in favor of pimpl.

I have a question if we should do it one by one, only when its needed or if we should add the pimpls to all API classes at once preemptively to break the ABI the least amount of times?

Comment (jan-kolarik) [reply include/libdnf5/base/goal_elements.hpp:174]: > I am also in favor of pimpl.
> 
> I have a question if we should do it one by one, only when its needed or if we should add the pimpls to all API classes at once preemptively to break the ABI the least amount of times?

I like the preemption idea. If we can detect all potential places now, I am for preparing the codebase for this purpose in advance.

File: dnf5/commands/history/history_store.cpp (hunk 1 above)
Comment (kontura) [reply dnf5/commands/history/history_store.cpp]: I have changed it, please take a look.

File: include/libdnf5/base/goal_elements.hpp (hunk 2 above)
Comment (kontura) [reply include/libdnf5/base/goal_elements.hpp]: (I have removed the problematic part from the PR so it doesn't block it, there is only a TODO for now.)

Comment (j-mracek) [reply include/libdnf5/base/goal_elements.hpp]: If we will create `pimpl`, how we can keep API compatibility? The structure does not have methods -> getters, setters

**Comment by j-mracek** [comment]
I am not sure whether `dnf history replay` is the best command. We can keep it as a compatibility alias, but I guess we could look for a better command. We already discussed to create `do` command to replace `shell`. What do you think?

File: include/libdnf5/base/transaction.hpp [hunk 3]
```diff
@@ -163,6 +163,10 @@ class Transaction {
     /// Retrieve a list of the problems that occurred during `check_gpg_signatures` procedure.
     std::vector<std::string> get_gpg_signature_problems() const noexcept;
 
+    /// Serialize the transaction into a json data format which can be later loaded
+    /// into a `libdnf5::Goal` and replayed.
+    std::string serialize_to_json();
```
Comment (j-mracek) [inline include/libdnf5/base/transaction.hpp:168]: The method is ok, but I have realized that there might be a white spot in our workflow if we want to download, test and store rpm path in the file.
Before download we can call `Command::goal_resolved()`, but there is no hook later.

Do you think whether it would be possible to return a different set of information if `serialize_to_json()`  as called before downloading of packages (without path to rpm) or after download (include paths of rpm)?

File: include/libdnf5/base/goal.hpp [hunk 4]
```diff
@@ -354,6 +354,12 @@ class Goal {
     void add_provide_install(
         const std::string & spec, const libdnf5::GoalJobSettings & settings = libdnf5::GoalJobSettings());
 
+    /// Add transaction replay request to the goal.
+    ///
+    /// @param json_serialized_transaction      A string of JSON serialized transaction
+    /// @param settings                         A structure to override default goal settings.
+    void add_transaction_replay(const std::string & json_serialized_transaction, libdnf5::GoalJobSettings & settings);
```
Comment (j-mracek) [inline include/libdnf5/base/goal.hpp:361]: What about to replace `replay` by `from_file`? The current PR allows not only replay the transaction but also add additional requests.

File: include/libdnf5/base/transaction.hpp (hunk 3 above)
Comment (kontura) [reply include/libdnf5/base/transaction.hpp:168]: > Do you think whether it would be possible to return a different set of information if `serialize_to_json()` as called before downloading of packages (without path to rpm) or after download (include paths of rpm)?

Yes, the `package_path` should be included only if the package is available locally.

File: include/libdnf5/base/goal.hpp (hunk 4 above)
Comment (kontura) [reply include/libdnf5/base/goal.hpp:361]: It is added from string so perhaps something like `add_transaction_actions_from_string()`?

Although I kind of like the `replay` word because it points to the format of the loaded string.

**Comment by kontura** [comment]
> I am not sure whether `dnf history replay` is the best command. We can keep it as a compatibility alias, but I guess we could look for a better command. We already discussed to create `do` command to replace `shell`. What do you think?

Sure, I think we will also need some command for the offline actions. I would do that in a different PR thought.
I can remove the `history replay` or we can keep it if it is going to stay as an alias.

**Comment by kontura** [comment]
I have marked the new API as `experimental` and also extended the `history store/replay` commands description with `[experimental]` keyword.

File: include/libdnf5/base/goal.hpp (hunk 4 above)
Comment (j-mracek) [reply include/libdnf5/base/goal.hpp:362]: The method that creates the string is called `serialize_to_json`. What about to add to the name both worlds. `add_serialized_transaction_from_json`  or shorter `add_serialized_transaction` with skipping type of information.
Anyway `add_transaction_actions_from_string()` is also ok.

File: dnf5/commands/history/history_store.cpp [hunk 5]
```diff
@@ -19,14 +19,75 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
...
+    int fd = mkstemp(dest.data());
+    if (fd == -1) {
+        throw std::filesystem::filesystem_error(
+            "cannot create temporary file", dest, std::error_code(errno, std::system_category()));
+    }
+
+    if (write(fd, json.data(), json.size()) == -1) {
+        throw std::filesystem::filesystem_error(
+            "cannot write to temporary file", dest, std::error_code(errno, std::system_category()));
+    }
```
Comment (j-mracek) [inline dnf5/commands/history/history_store.cpp:82]: I believe that the code will keep open the file descriptor and also the temp file will remain on the system. What about following code?
```
close(fd);
std::error_code ec;
std::filesystem::remove(tmpfilepath, ec);
```

File: dnf5/commands/history/history_store.cpp [hunk 6]
```diff
@@ -19,14 +19,75 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
...
+
+    if (close(fd) != 0) {
+        throw std::filesystem::filesystem_error(
```
Comment (j-mracek) [inline dnf5/commands/history/history_store.cpp:85]: The TMP file will remain on the system

File: dnf5/commands/history/history_store.cpp [hunk 7]
```diff
@@ -19,14 +19,75 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
...
+            "cannot close temporary file", dest, std::error_code(errno, std::system_category()));
+    }
+
+    //TODO(amatej): add warning if file already exists?
+    std::filesystem::rename(dest, output_option->get_value());
```
Comment (j-mracek) [inline dnf5/commands/history/history_store.cpp:90]: This might also fail - raises an error and tmp file will remain. I checked our code and we never handle that situation.

File: libdnf5/base/goal.cpp [hunk 8]
```diff
@@ -203,6 +205,9 @@ class Goal::Impl {
 
     void install_group_package(base::Transaction & transaction, libdnf5::comps::Package pkg);
     void remove_group_packages(const rpm::PackageSet & remove_candidates);
+
+    transaction::TransactionReplay transaction_replay;
+    GoalJobSettings transaction_replay_settings;
```
Comment (j-mracek) [inline libdnf5/base/goal.cpp:210]: What about to replace it by a vector with pair type inside. An empty vector should be cheaper then two default structures. What do you think?

File: libdnf5/base/goal.cpp [hunk 9]
```diff
@@ -2252,11 +2396,21 @@ base::Transaction Goal::resolve() {
...
+
     transaction.p_impl->set_transaction(p_impl->rpm_goal, module_sack, ret);
 
     return transaction;
 }
 
+
+void Goal::add_transaction_replay(
+    const std::string & json_serialized_transaction, libdnf5::GoalJobSettings & settings) {
+    //TODO(amatej): allow only one replay per goal?
```
Comment (j-mracek) [inline libdnf5/base/goal.cpp:2409]: I guess that we should add a limitation (exception)  or support more transactions. The current implementation performs the last provided transaction, and I don't think that this is right.
If we will add support only for one file per goal we can anytime extend the support to multi transactions support without changing API, ABI, or introducing a new exception. If we will start with a support of multiple files then adding the restriction might be painful. May I ask you for your preferences?

File: libdnf5/base/transaction.cpp [hunk 10]
```diff
@@ -448,6 +449,22 @@ void Transaction::Impl::set_transaction(
...
         packages.emplace_back(std::move(tspkg));
     }
+
+    // After all packages were added check rpm reason overrides
+    //TODO(amatej): Should these reason overrides be done via a REASON_CHANGE action? This would show up in the transaction table
+    if (!rpm_reason_overrides.empty()) {
+        for (auto & pkg : packages) {
+            const auto reason_override = rpm_reason_overrides.find(pkg.get_package().get_nevra());
+            if (reason_override != rpm_reason_overrides.end()) {
+                // For all other actions we don't want to lower the reason
```
Comment (j-mracek) [inline libdnf5/base/transaction.cpp:459]: I am not sure about the condition, may be after finishing the review I will have a better picture.

Comment (j-mracek) [reply libdnf5/base/transaction.cpp:459]: I will test some edge cases

**Comment by j-mracek** [comment]
I discovered confusing output
I did `dnf5 install acpi`   then `dnf5 mark weak acpi`, then stored the transaction with reason change, then `dnf5 history replay transaction.json`
```
dnf5 history replay transaction.json
Updating and loading repositories:
Repositories loaded.
Failed to resolve the transaction:
Argument 'acpi-0:1.7-20.fc38.x86_64' matches only excluded packages.
```
`acpi` was not excluded, because I was able to install it.

**Comment by j-mracek** [comment]
@kontura May I ask you to verify following scenario
`sudo dnf5 history store 57`
`dnf5 history store 56`
Expected fail - unable to replace the file

**Comment by j-mracek** [comment]
Similar to https://github.com/rpm-software-management/dnf5/pull/1005#issuecomment-1824327195
When I have installed a different version when transaction is a reason change.
```
sudo dnf5 history replay transaction.json
Updating and loading repositories:
Repositories loaded.
Failed to resolve the transaction:
Argument 'iftop-0:1.0-0.30.pre4.fc38.x86_64' matches only excluded packages.
```

File: libdnf5/base/goal.cpp [hunk 11]
```diff
@@ -596,6 +601,130 @@ GoalProblem Goal::Impl::add_module_specs_to_goal(base::Transaction & transaction
     return ret;
 }
 
+GoalProblem Goal::Impl::add_transaction_replay_specs_to_goal(base::Transaction & transaction) {
+    libdnf5::GoalJobSettings settings = transaction_replay_settings;
+    bool skip_unavailable = settings.resolve_skip_unavailable(base->get_config());
+
+    for (const auto & package_replay : transaction_replay.packages) {
+        libdnf5::GoalJobSettings settings_per_package = settings;
```
Comment (j-mracek) [inline libdnf5/base/goal.cpp:609]: May I ask you why you create a copy of settings here?

File: libdnf5/base/goal.cpp [hunk 12]
```diff
@@ -596,6 +601,130 @@ GoalProblem Goal::Impl::add_module_specs_to_goal(base::Transaction & transaction
...
+            transaction.p_impl->rpm_reason_overrides[package_replay.nevra] = package_replay.reason;
+        } else if (package_replay.action == transaction::TransactionItemAction::REMOVE) {
+            rpm_specs.emplace_back(GoalAction::REMOVE, package_replay.nevra, settings_per_package);
+            transaction.p_impl->rpm_reason_overrides[package_replay.nevra] = package_replay.reason;
+        } else if (package_replay.action == transaction::TransactionItemAction::REPLACED) {
+            if (!skip_unavailable) {
+                rpm_specs.emplace_back(GoalAction::REMOVE, package_replay.nevra, settings_per_package);
+            }
+        } else if (package_replay.action == transaction::TransactionItemAction::REASON_CHANGE) {
+            rpm_reason_change_specs.emplace_back(
```
Comment (j-mracek) [inline libdnf5/base/goal.cpp:644]: Shell we pass full NEVRA here or only name.arch of the package?

File: libdnf5/base/transaction.cpp (hunk 10 above)
Comment (j-mracek) [reply libdnf5/base/transaction.cpp:459]: `// For all other actions we don't want to lower the reason`
May I ask you why?

What do you think  is more important  -  reason that is listed in transaction file or reason in system-state (internal DNF5 DB)?

Note:
Upgrade is an operation that normally does not change reason,  right? But when upgrading  package also obsoletes another package than the final reason is the strongest of both.

File: include/libdnf5/base/goal.hpp (hunk 4 above)
Comment (kontura) [reply include/libdnf5/base/goal.hpp:362]: I like the shorter `add_serialized_transaction()` which drops the JSON. I am thinking about dropping it also from the `serialize_to_json()`.
So that we have just:
- `libdnf5::base::transaction::serialize()`
- `libdnf5::transaction::transaction::serialize()`
- `libdnf5::goal::add_serialized_transaction()`

File: libdnf5/base/goal.cpp (hunk 9 above)
Comment (kontura) [reply libdnf5/base/goal.cpp:2409]: I think limiting to one is the right approach. You are right we can always extend it later if there is a use case.

File: dnf5/commands/history/history_store.cpp (hunk 5 above)
Comment (kontura) [reply dnf5/commands/history/history_store.cpp:82]: I think it would be best to use the `libdnf5::utils::fs::temp.hpp` to handle this. I made a PR to expose it: https://github.com/rpm-software-management/dnf5/pull/1050

File: libdnf5/base/goal.cpp (hunk 8 above)
Comment (kontura) [reply libdnf5/base/goal.cpp:210]: I think vector might be a little bit confusing if we only allow one replay per goal. I will try `std::unique_ptr`.

**Comment by kontura** [comment]
> I discovered confusing output I did `dnf5 install acpi` then `dnf5 mark weak acpi`, then stored the transaction with reason change, then `dnf5 history replay transaction.json`
> 
> ```
> dnf5 history replay transaction.json
> Updating and loading repositories:
> Repositories loaded.
> Failed to resolve the transaction:
> Argument 'acpi-0:1.7-20.fc38.x86_64' matches only excluded packages.
> ```
> 
> `acpi` was not excluded, because I was able to install it.

I am not able to reproduce this, for me it says: 
```
Updating and loading repositories:
Repositories loaded.
Package "acpi-1.7-21.fc39.x86_64" is already installed with reason "Weak Dependency".

Nothing to do.
```
Which seems reasonable. 
Is there some step missing in the reproducer?

File: libdnf5/base/goal.cpp (hunk 11 above)
Comment (kontura) [reply libdnf5/base/goal.cpp:609]: The `settings` are used to select the repository which can be different for each action. I wanted to avoid the actions affecting each other.
But then it is copied again into the particular vector of specs so it would be possible to have the settings just once and make sure we always overwrite the `to_repo_ids` or other used values.

I can change it if you prefer that? It would be more efficient.

File: libdnf5/base/goal.cpp (hunk 12 above)
Comment (kontura) [reply libdnf5/base/goal.cpp:644]: It would make the `REASON_CHANGE` work even if a different version than expected is installed.
But why should it work differently than all other actions?

File: libdnf5/base/transaction.cpp (hunk 10 above)
Comment (kontura) [reply libdnf5/base/transaction.cpp:459]: I think the more important reason is the stronger one. Just like with the obsoletes.

Specifically there is a [test scenario](https://github.com/rpm-software-management/ci-dnf-stack/blob/main/dnf-behave-tests/dnf/transaction-sr/replay.feature#L1228-L1282) where the replayed transaction has an upgrade with `Dependency` but on host the package has `User` reason. I think it would be wrong to lower the reason to `Dependency`. The package is user-installed.

So I think we don't want to lower the reason because some user installed packages could get marked as dependencies.

File: libdnf5/base/goal.cpp (hunk 11 above)
Comment (j-mracek) [reply libdnf5/base/goal.cpp:611]: Thank you for explanation. I am fine with that.

**Comment by j-mracek** [comment]
> > I discovered confusing output I did `dnf5 install acpi` then `dnf5 mark weak acpi`, then stored the transaction with reason change, then `dnf5 history replay transaction.json`
> > ```
> > dnf5 history replay transaction.json
> > Updating and loading repositories:
> > Repositories loaded.
> > Failed to resolve the transaction:
> > Argument 'acpi-0:1.7-20.fc38.x86_64' matches only excluded packages.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `acpi` was not excluded, because I was able to install it.
> 
> I am not able to reproduce this, for me it says:
> 
> ```
> Updating and loading repositories:
> Repositories loaded.
> Package "acpi-1.7-21.fc39.x86_64" is already installed with reason "Weak Dependency".
> 
> Nothing to do.
> ```
> 
> Which seems reasonable. Is there some step missing in the reproducer?

The problem was fixed just recently (After my testing) - see https://github.com/rpm-software-management/dnf5/commit/763f9ccdfccc3c970603da1f7f5d9ebabd861b08

File: libdnf5/base/goal.cpp (hunk 12 above)
Comment (j-mracek) [reply libdnf5/base/goal.cpp:646]: This is a good question with no good answer.

I can argue that:
`dnf5 mark` with argument `dnf` will be impossible to replay when a different version is installed. We also know, that internal implementation does not care about version and only stores `name.arch` information.
In the following example the original version was not in any repository

```
$ sudo dnf5 mark dependency yum
$ dnf5 history store 61
$ sudo dnf5 upgrade yum
$ sudo dnf5 mark user  yum
$ dnf5 history replay transaction.json
Updating and loading repositories:
Repositories loaded.
Failed to resolve the transaction:
No match for argument: yum-0:4.18.1-20231110004827.0.g41a287e2.fc38.noarch
```

When I replay reason change (to dep) after I changed installed version and also after I changed reason (from dep to user) I got following problem.

```
dnf5 history replay transaction.json
Updating and loading repositories:
Repositories loaded.
Failed to resolve the transaction:
Argument 'yum-0:4.18.1-20231130005028.8.g6229c1d4.fc38.noarch' matches only excluded packages.
```

I think that the implementation in PR is more strict then it has to be. Then it is not robust to any changes of installed package when the reason change is replayed.

File: libdnf5/base/goal.cpp (hunk 8 above)
Comment (j-mracek) [reply libdnf5/base/goal.cpp]: Perfect.

File: libdnf5/base/transaction.cpp (hunk 10 above)
Comment (j-mracek) [reply libdnf5/base/transaction.cpp]: Thank you for the explanation. Looks good to me

**Comment by j-mracek** [comment]
I think there are 3 remaining items:

- Behavior related to reason change
- name of command `history replay`
- allow silent file override for `transaction.json`
  - Personally I have no problem with that, but it would be good that DNF5 will behave uniformly in those situations

I consider only `- name of command 'history replay'` as a item that we should address in this PR.

**Comment by j-mracek** [comment]
LGTM
